File: week 2/p2 포커 패 (Poker Hands)/Main.py
value_dict = {'T':10, 'J':11, 'Q':12, 'K':13, 'A':14}

def poker(card):

    values = sorted([c[0] for c in card])
    pattern = [c[1] for c in card]
    straight = (values == list(range(min(values), max(values)+1)))
    flush = all(p == pattern[0] for p in pattern)

    if straight and flush:
        return 8, max(values), values

    pairs = []
    three_card = False
    pair_present = False
    three_card_value = -1

    for v in set(values):
        # four card
        if values.count(v) == 4:
            return 7, v, values
        if values.count(v) == 3:
            three_card = True
            three_card_value = v
        if values.count(v) == 2:
            pair_present = True
            pairs.append(v)

    if three_card and pair_present:
        return 6, three_card_value, values
    if flush:
        return 5, max(values), values
    if straight:
        return 4, max(values), values
    if three_card:
        values.remove(three_card_value)
        return 3, three_card_value, values
    if len(pairs) == 2:
        pair1 = pairs[0]
        pair2 = pairs[1]

        values.remove(pair1)
        values.remove(pair2)
        values.remove(pair1)
        values.remove(pair2)

        if pair1>=pair2:
            return 2, pair1, [pair2, values[0]]
        elif pair1<pair2:
            return 2, pair2, [pair1, values[0]]
    if len(pairs) == 1:
        values.remove(pairs[0])
        values.remove(pairs[0])
        return 1, pairs[0], values


    return 0, max(values), values


def checkHighCard(black_card, white_card):
    for i in range(len(black_card)).__reversed__():
        if black_card[i] == white_card[i]:
            continue
        elif black_card[i] > white_card[i]:
            print("Black wins.")
            return
        elif black_card[i] < white_card[i]:
            print('White wins.')
            return

    print('Tie.')


while True:
    try:
        # 카드 input받기
        temp = input().split()
        numbering = []

        # T, K, J, Q를 정수로 바꾼 후, 카드 타입과 함께 튜플로 저장
        for card in temp:
            if card[0] not in value_dict:
                value = int(card[0])
            else:
                value = value_dict[card[0]]
            numbering.append((value, card[1]))

        black_card = numbering[:5]
        white_card = numbering[5:]

        # poker check
        blackClass, blackNum, bRemainderCard = poker(black_card)
        whiteClass, whiteNum, wReaminderCard = poker(white_card)

        if blackClass > whiteClass:
            print("Black wins.")
        elif whiteClass > blackClass:
            print("White wins.")
        else:
            # high card같은 경우는 따로 계산
            if blackNum > whiteNum:
                print("Black wins.")
            elif whiteNum > blackNum:
                print("White wins.")
            else:
                if blackClass < 3:
                    checkHighCard(bRemainderCard, wReaminderCard)
                else:
                    print('Tie.')
    except:
        break

File: week 2/p2 포커 패 (Poker Hands)/Main.c
#include <stdio.h>
#define get_value(x) ((x) / 10)
#define get_suit(x) ((x) % 10)

/* 문자열 형태의 카드 값을 숫자 표현으로 변환 */
int encode_card(char *card)
{
	int result;
	
	switch (card[0]) {
		case 'T': result = 100; break;
		case 'J': result = 110; break;
		case 'Q': result = 120; break;
		case 'K': result = 130; break;
		case 'A': result = 140; break;
		default : result = (card[0] - '0') * 10;
	}
	switch (card[1]) {
		case 'H': result += 1; break;
		case 'D': result += 2; break;
		case 'S': result += 3; break;
		case 'C': result += 4; break;
	}
	return result;
}

/* ranking function */
long get_hand_value(int hand[5])
{
	int i, j, max, temp;
	int value[5], suit[5];
	long result;
	
	/* 패를 정렬해두면 족보 판정을 훨씬 쉽게 할 수 있고 같은 족보일 때 비교하기 위한
	제일 높은 카드도 쉽게 찾을 수 있으므로 hand value를 쉽게 계산할 수 있다. */
	for (i = 0; i < 4; i++) {
		max = i;
		for (j = i + 1; j < 5; j++)
			if (hand[j] > hand[max])
				max = j;
		temp = hand[i];
		hand[i] = hand[max];
		hand[max] = temp;
	}
	
	for (i = 0; i < 5; i++) {
		value[i] = get_value(hand[i]);
		suit[i] = get_suit(hand[i]);
	}
	
	/* straight flush */
	if (value[1] + 1 == value[0] && suit[1] == suit[0]
		 && value[2] + 2 == value[0] && suit[2] == suit[0]
		 && value[3] + 3 == value[0] && suit[3] == suit[0]
		 && value[4] + 4 == value[0] && suit[4] == suit[0])
		
		result = (9 << 20) + (value[0] << 16);
	
	/* four of a kind 정렬된 상태이므로 카드를 일일히 세지 않고도
	1, 4번째 카드가 같거나 또는 2, 5번째 카드가 같은지만 검사하면 된다. */
	else if (value[0] == value[3] || value[1] == value[4])
		
		result = (8 << 20) + (value[1] << 16); /* 2번째 카드는 무조건 포카드에 포함된다 */
	
	/* full house. 2장+3장 또는 3장+2장 형태이므로
	다음과 같이 아주 간단하게 검사할 수 있다. */
	else if (value[0] == value[2] && value[3] == value[4]) /* 3장 + 2장 */
		
		result = (7 << 20) + (value[0] << 16);
	
	else if (value[0] == value[1] && value[2] == value[4]) /* 2장 + 3장 */
		
		result = (7 << 20) + (value[2] << 16);
	
	/* flush */
	else if (suit[1] == suit[0] && suit[2] == suit[0]
					&& suit[3] == suit[0] && suit[4] == suit[0])
		
		result = (6 << 20) + (value[0] << 16) + (value[1] << 12)
		+ (value[2] << 8) + (value[3] << 4) + value[4];
	
	/* straight */
	else if (value[1] + 1 == value[0] && value[2] + 2 == value[0]
					&& value[3] + 3 == value[0] && value[4] + 4 == value[0])
		
		result = (5 << 20) + (value[0] << 16);
	
	/* three of a kind */
	else if (value[0] == value[2] || value[1] == value[3]
					|| value[2] == value[4])
		
		result = (4 << 20) + (value[2] << 16); /* 3번째 카드는 무조건 쓰리카드에 포함된다! */
	
	/* two pairs 다음과 같은 세가지 가능성만 존재한다. */
	else if (value[0] == value[1] && value[2] == value[3])
		
		result = (3 << 20) + (value[1] << 16) + (value[3] << 12) + (value[4] << 8);
	
	else if (value[0] == value[1] && value[3] == value[4])
		
		result = (3 << 20) + (value[1] << 16) + (value[3] << 12) + (value[2] << 8);
	
	else if (value[1] == value[2] && value[3] == value[4])
		
		result = (3 << 20) + (value[1] << 16) + (value[3] << 12) + (value[0] << 8);
	
	/* pare */
	else if (value[0] == value[1])
		
		result = (2 << 20) + (value[0] << 16) + (value[2] << 12) + (value[3] << 8) + (value[4] << 4);
	
	else if (value[1] == value[2])
		
		result = (2 << 20) + (value[1] << 16) + (value[0] << 12) + (value[3] << 8) + (value[4] << 4);
	
	else if (value[2] == value[3])
		
		result = (2 << 20) + (value[2] << 16) + (value[0] << 12) + (value[3] << 8) + (value[4] << 4);
	
	else if (value[3] == value[4])
		
		result = (2 << 20) + (value[3] << 16) + (value[0] << 12) + (value[3] << 8) + (value[4] << 4);
	
	/* high card */
	else
		result = (1 << 20) + (value[0] << 16) + (value[1] << 12) + (value[2] << 8)
		+ (value[3] << 4) + value[4];
	
	return result;
}

void main(void)
{
	char line[100];
	int hand[2][5];
	long hand_value[2];
	int i, j, t;
	
	while (gets(line) && *line) {
		t = 0;
		
		for (i = 0; i < 2; i++) {
			for (j = 0; j < 5; j++) {
				while (line[t] == ' ')
					t++;
				hand[i][j] = encode_card(line + t);
				t += 2;
			}
			hand_value[i] = get_hand_value(hand[i]);
		}
		
		if (hand_value[0] > hand_value[1])
			puts("Black wins.");
		else if (hand_value[0] < hand_value[1])
			puts("White wins.");
		else
			puts("Tie.");
	}
}














